src/transformers/models/bloom/modeling_bloom.py

```diff
@@ -231,105 +228,129 @@ def __init__(self, config, layer_number=None):
+        key_layer = key_layer.permute(0, 2, 3, 1).reshape(batch_size * self.num_heads, self.head_dim, q_length)
+        value_layer = value_layer.transpose(1, 2).reshape(batch_size * self.num_heads, q_length, self.head_dim)
         if layer_past is not None:
             past_key, past_value = layer_past
-            # concatenate along seq_length dimension -> [batch_size, qk_length, num_heads, head_dim]
-            key_layer = torch.cat((past_key.type_as(key_layer), key_layer), dim=1)
-            value_layer = torch.cat((past_value.type_as(value_layer), value_layer), dim=1)
+            # concatenate along seq_length dimension:
+            #  - key: [batch_size * self.num_heads, head_dim, kv_length]
+            #  - value: [batch_size * self.num_heads, kv_length, head_dim]
+            key_layer = torch.cat((past_key, key_layer), dim=2)
+            value_layer = torch.cat((past_value, value_layer), dim=1)
+
+        _, _, kv_length = key_layer.shape
 
         if use_cache is True:
             present = (key_layer, value_layer)
         else:
             present = None
 
-        beta = 1.0 / self.layer_number
-
-        # # [batch_size*num_heads, head_dim, q_length] x [batch_size*num_heads, head_dim, k_length] -> [batch_size*num_heads, q_length, k_length]
-        matmul_result = (1.0 / self.norm_factor) * torch.bmm(
-            query_layer.transpose(1, 2).reshape(-1, query_layer.shape[1], query_layer.shape[3]),
-            key_layer.permute(0, 2, 3, 1).reshape(-1, key_layer.shape[3], key_layer.shape[1]),
-        ) + beta * alibi
+        # [batch_size * num_heads, q_length, kv_length]
+        matmul_result = torch.baddbmm(
+            input=alibi,
+            batch1=query_layer,
+            batch2=key_layer,
+            beta=self.beta,
+            alpha=self.inv_norm_factor,
+        )
 
-        # change view to [batch_size, num_heads, q_length, k_length]
-        attention_scores = matmul_result.view(-1, self.num_heads, matmul_result.size(1), matmul_result.size(2))
+        # change view to [batch_size, num_heads, q_length, kv_length]
+        attention_scores = matmul_result.view(batch_size, self.num_heads, q_length, kv_length)
 
-        # We replace the scaled softmax by just a few line of code - [batch_size, num_heads, q_length, k_length]
+        # we cast attention scores to fp32, compute scaled softmax and cast back into initial dtype - [batch_size, num_heads, q_length, kv_length]
         input_dtype = attention_scores.dtype
-        attn_weights = (attention_scores * self.layer_number) + attention_mask
-        attn_weights = torch.max(attn_weights, torch.tensor(torch.finfo(attn_weights.dtype).min))
-        attention_probs = nn.functional.softmax(attn_weights, dim=-1, dtype=torch.float32).to(input_dtype)
-        attention_probs = attention_probs * (~attention_mask.bool())
-        # [batch_size, num_heads, q_length, k_length]
+        attention_scores = attention_scores.float()
```

Comment:
I'm thinking that if `dtype=torch.float16` then we should cast, but not if `dtype=torch.bfloat16`. cc @younesbelkada

Reply:
I think that we always have to cast to fp32 - at least that is what I have understood from the Meg-DS implementation: https://github.com/bigscience-workshop/Megatron-DeepSpeed/blob/0f23a729ec5ed6ef46e1fecc8071ac11acdb91c6/megatron/model/fused_softmax.py#L212 + the logits tests of the small model that you can find on the testing file for the fused kernel test.

Reply:
In the cuda kernel it seems the values are stored in fp32: https://github.com/bigscience-workshop/Megatron-DeepSpeed/blob/0f23a729ec5ed6ef46e1fecc8071ac11acdb91c6/megatron/fused_kernels/scaled_masked_softmax.h#L318

Reply:
But would love to have your thoughts on that, maybe I am wrong!

Reply:
I think the `layer_number` scaling was only useful for float16, and we should not need it anymore (remember when we tried training 100B on float16). Also the current status in `main` is to only use `softmax(dtype=torch.float)` which is the same behaviour as here (besides the fact that I've removed the `layer_number` scaling trick).

Reply:
Actually I think this is good, I still have a small doubt since the values of `torch.finfo(torch.float32).min` and `torch.finfo(torch.bfloat16).min` are not the same (with the fp32 one being larger). So maybe as a sanity check it would be nice to run some batched generation to make sure it works in bf16

Reply:
I really don't think it matters as after the softmax those values are going to clearly be 0, unless an activation is in the same order of magnitude of `-1e+38` (which to be honest is a much larger problem if that's true)